- update call-x-callAsync.md docs to mirror our API (#13198)
meteor/meteor: v3-docs/v3-migration-docs/breaking-changes/call-x-callAsync.md

# Meteor.call x Meteor.callAsync

::: tip

Use `Meteor.callAsync` instead of `Meteor.call` when handling async methods. If you use `Meteor.call` with
async methods it can lead to unexpected behavior.

Using `Meteor.callAsync` will make your code more predictable and easier to maintain.

:::

Example of how to migrate from `Meteor.call` to `Meteor.callAsync`:

::: code-group

```js [v2-client.jsx]
import { Meteor } from "meteor/meteor";

let data, error;

> Meteor.call("getAllData", (err, res) => {
>   // [!code highlight]

Reviewer: ```suggestion
Meteor.call("getAllData", (err, res) => {   // [!code highlight]
```

Reviewer 2: @denihs does this make sense? Why is it? @henriquealbert

# Can I still use Meteor.call?

You can, but we recommend you use it only to call the sync method.

In fact, we log a warning message if you use `Meteor.call` to call an async method.

If you use `Meteor.call` with async methods it can lead to unexpected behavior.

> `Meteor.callAsync` is the standard for calling async methods.

Reviewer 3: It might be more accurate as something like:

```
You can, but we recommend you use it only to call methods that do not have a method stub, or when the method stub is synchronous.

In fact, we log a warning message if you use `Meteor.call` to call a method with an async method stub since it can lead to unexpected behavior.

 `Meteor.callAsync` is the standard for calling methods and supports any method, including those that have an async method stub.
```

It might also be good to have a reminder of what a method stub is.

> Here is also important to remember what a stub is. A stub is a client-side simulation of the server-side method that runs immediately when the method is invoked, allowing the client to update its state optimistically before receiving the server's response. So, basically any MongoDB function (insertAsync, updateAsync, etc...) that's called inside a method it's considered a stub.

Reviewer 3: ```suggestion
Here is also important to remember what a stub is. A stub is a client-side simulation of the server-side method that runs immediately when the method is invoked, allowing the client to update its state optimistically before receiving the server's response. So, basically any Meteor method that is defined on the client is considered a stub.
```

  getAllData() {
    return MyCollection.find().fetch(); // [!code highlight]
  },
});
```

```js [v3-client.jsx]
import { Meteor } from "meteor/meteor";

try {
  const data = await Meteor.callAsync("getAllData"); // [!code highlight]
  // render data
} catch (error) {
  // render error
}
```

```js [v3-server.js]
import { Meteor } from "meteor/meteor";
import { Mongo } from "meteor/mongo";

const MyCollection = new Mongo.Collection("myCollection");

Meteor.methods({
  async getAllData() {
    return await MyCollection.find().fetchAsync(); // [!code highlight]
  },
});
```

:::

## Rules of using `Meteor.callAsync` & `Meteor.call`

You can still use `Meteor.callAsync` and `Meteor.call` together. But there are some catches that you need to
pay attention to. Let's go over them.

### How to NOT use them together

These are for **async** method when called with `Meteor.callAsync` or `Meteor.call`.

**Case 1**: Calling `Meteor.callAsync` and then right after calling `Meteor.call` when the methods have stubs.

```js
import { Meteor } from "meteor/meteor";

Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB")
  .then((data) => console.log(data))
  .catch((err) => console.log(err));

> Meteor.call("SOME_ASYNC_METHOD_WITH_STUB", (err, res) => {

Reviewer 3: I'm assuming the stub used with `Meteor.call` is async. This seems to be broken due to the changes in `EV.withValue` to support promises on the client: https://github.com/meteor/meteor/blob/818c5da95ca5b884e61ad153ad27ccc7ec86741b/packages/meteor/dynamics_browser.js#L46-L52

Before, async stubs were not supported in `Meteor.apply`. With this change, they still aren't really supported, but it introduces the same issues we had with `Meteor.callAsync` in Meteor 2 that the queueing fixed.

 If we do want to support them, [this code](https://github.com/meteor/meteor/blob/818c5da95ca5b884e61ad153ad27ccc7ec86741b/packages/ddp-client/common/livedata_connection.js#L612-L620) would need to be updated, as well as how the queueing handles `Meteor.apply`.

Though I don't think we should support async stubs in `Meteor.apply`.

  if (err) {
    console.log(err);
  } else {
    console.log(res);
  }
});
```

> **Case 2**: Calling `Meteor.call` and then right after calling `Meteor.callAsync` when the methods have stubs.

Reviewer 3: Is the stub used with `Meteor.call` here async? If it is, this is the same issue as with case 1 - `Meteor.call` needs to either not be used with async stubs, or should be fixed so it queues correctly with async stubs.

```js
import { Meteor } from "meteor/meteor";

Meteor.call("SOME_ASYNC_METHOD_WITH_STUB", (err, res) => {
  if (err) {
    console.log(err);
  } else {
    console.log(res);
  }
});

Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB")
  .then((data) => console.log(data))
  .catch((err) => console.log(err));
```

**Case 3**: Calling `Meteor.callAsync` and then right after calling `Meteor.call` but the method called with `Meteor.call`
does not have stub.

```js
import { Meteor } from "meteor/meteor";

Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB")
  .then((data) => console.log(data))
  .catch((err) => console.log(err));

Meteor.call("SOME_ASYNC_METHOD_WITHOUT_STUB", (err, res) => {
  if (err) {
    console.log(err);
  } else {
    console.log(res);
  }
});
```

> **Case 4**: Calling `Meteor.call` and then right after calling `Meteor.callAsync` but the method called with `Meteor.call`

Reviewer 3: What is the problem with case 4? This should be supported.

does not have stub.

```js
import { Meteor } from "meteor/meteor";

> Meteor.call("SOME_ASYNC_METHOD_WITHOUT_STUB", (err, res) => {

Reviewer 3: I think using `ASYNC_METHOD` makes this a little confusing. The client isn't affected by if the method is async or not on the server. It might be better to instead say if the stub is async or sync, since that is what affects the client.

specific scenario, use it only when calling sync methods.

Here are some examples:

```js
import { Meteor } from "meteor/meteor";

// [!code highlight] Example 1: Call serveral Meteor.callAsync in a row.
// [!code highlight] It doesn't matter if the method has stub or not.

Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB");
Meteor.callAsync("SOME_ASYNC_METHOD_WITHOUT_STUB");
Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB");
Promise.all([
  Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB"),
  Meteor.callAsync("SOME_ASYNC_METHOD_WITHOUT_STUB"),
  Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB"),
]).then(([data1, data2]) => {
  console.log(data1, data2);
});

// [!code highlight] Example 2: Call several async methods with `Meteor.callAsync`
// [!code highlight] and several sync methods with `Meteor.call` in row

Meteor.call("SOME_SYNC_METHOD");
Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB");
Meteor.call("SOME_SYNC_METHOD");
Meteor.callAsync("SOME_ASYNC_METHOD_WITH_STUB");
Meteor.callAsync("SOME_ASYNC_METHOD_WITHOUT_STUB");
```

As you can see, if you follow the rule of calling async methods with `Meteor.callAsync` and sync methods with 

> `Meteor.call` you can use them as you want. But ideally, after you finish your migration, you should have any `Meteor.call`

Reviewer 4: > ideally, after you finish your migration, you should have any `Meteor.call`

Shouldn't `should` be `shouldn't` here? :) 

Also, I suggest putting this near the top of the "Rules of using Meteor.callAsync & Meteor.call" so it reads something like:

> Technically, `Meteor.callAsync` and `Meteor.call` can be used alongside each other. But ideally, after you finish your migration, you shouldn't use `Meteor.call`, so you don't need to worry about the gotcha scenarios detailed below.

});

// render data or error
```

```js [v2-server.js]
import { Meteor } from "meteor/meteor";
import { Mongo } from "meteor/mongo";

const MyCollection = new Mongo.Collection("myCollection");

Meteor.methods({
  getAllData() {
    return MyCollection.find().fetch(); // [!code highlight]
  },
});
```

```js [v3-client.jsx]
import { Meteor } from "meteor/meteor";

try {
  const data = await Meteor.callAsync("getAllData"); // [!code highlight]
  // render data
} catch (error) {
  // render error
}
```

```js [v3-server.js]
import { Meteor } from "meteor/meteor";
import { Mongo } from "meteor/mongo";

const MyCollection = new Mongo.Collection("myCollection");

Meteor.methods({
  async getAllData() {
    return await MyCollection.find().fetchAsync(); // [!code highlight]
  },
});
```

:::

> # What are the limitations of call Meteor.callAsync

Reviewer 3: ```suggestion
# What are the limitations of Meteor.callAsync
```

When we introduced [async Method stubs](https://guide.meteor.com/2.8-migration.html#callasync) the implementation brought some [limitations](https://github.com/zodern/fix-async-stubs#:~:text=Specifically%2C%20while%20an,used%20in%20stubs).

Those limitations were addressed in this [package](https://github.com/zodern/fix-async-stubs/) created by [Zodern](https://github.com/zodern), and later, we moved the solution to the [core](https://github.com/meteor/meteor/blob/ecdfd3c610fbe5334eee024702fe0c354944f58b/packages/ddp-client/client/queueStubsHelpers.js). 

But there is no perfect solution to the problems with async stubs.

To ensure other code will not run while an async stub is running, async stubs can not use these API's:

- fetch/XMLHttpRequest
- setTimeout or setImmediate
- indexedDB
- web workers

> - any other web api that schedules macrotasks

Reviewer 3: I'm not sure if this is any better:

```suggestion
- any other async web api's that wait on macrotasks
```

> - Using these API's could allow other code to run before the async stub finishes.

Reviewer 3: ```suggestion

Using these API's could allow other code to run before the async stub finishes.
```
